From 870897dde20242ef80b7426755beebc79df28a43 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michel=20D=C3=A4nzer?= <mdaenzer@redhat.com>
Date: Tue, 1 Sep 2026 17:42:02 +0200
Subject: [PATCH] onscreen/native: Disable FBO path by default with software
 rendering

It results in a blank screen for me with llvmpipe on bare metal, and
in glitches on the VM setup in the issue below.

Closes: https://gitlab.gnome.org/GNOME/mutter/-/work_items/4998
---
 src/backends/native/meta-onscreen-native.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/src/backends/native/meta-onscreen-native.c b/src/backends/native/meta-onscreen-native.c
index 32aa70896f..0f9bbcc501 100644
--- a/src/backends/native/meta-onscreen-native.c
+++ b/src/backends/native/meta-onscreen-native.c
@@ -3232,8 +3232,8 @@ should_try_fbos (CoglOnscreen *onscreen)
   CoglFramebuffer *framebuffer = COGL_FRAMEBUFFER (onscreen_native);
   CoglContext *cogl_context = cogl_framebuffer_get_context (framebuffer);
   CoglDisplay *cogl_display = cogl_context_get_display (cogl_context);
-  CoglRendererEGL *renderer_egl =
-    COGL_RENDERER_EGL (cogl_display_get_renderer (cogl_display));
+  CoglRenderer *renderer = cogl_display_get_renderer (cogl_display);
+  CoglRendererEGL *renderer_egl = COGL_RENDERER_EGL (renderer);
   gboolean is_nvidia, has_partial_updates;
   const char *use_fbos;
 
@@ -3250,6 +3250,14 @@ should_try_fbos (CoglOnscreen *onscreen)
       return try;
     }
 
+  if (!cogl_renderer_is_hardware_accelerated (renderer))
+    {
+      meta_topic (META_DEBUG_KMS,
+                  "Not trying FBO path because "
+                  "renderer isn't hardware accelerated");
+      return FALSE;
+    }
+
   is_nvidia = meta_onscreen_native_is_nvidia (onscreen_native);
 
   /* The Mesa Gallium HUD can only work with a surface */
-- 
2.55.0

